feat: Support Moore Threads GPU

yeahdongcn opened the pull request:
[Moore Threads](https://en.mthreads.com/), a cutting-edge GPU startup, introduces MUSA (Moore Threads Unified System Architecture) as its foundational technology. This pull request marks the initial integration of MTGPU support into llama.cpp, leveraging MUSA's capabilities to enhance LLM inference performance.

Similar to https://github.com/ggerganov/llama.cpp/pull/1087, CUDA APIs are replaced by MUSA APIs using macros, and a new build option is added to Makefile and CMake.

```bash
# make
make GGML_MUSA=1

# CMake
cmake -B build -DGGML_MUSA=ON
cmake --build build --config Release
```

I also sent a [PR](https://github.com/ollama/ollama/pull/5556) to [Ollama](https://github.com/ollama/ollama) to integrate MTGPU to it and all the tests were performed through [Ollama](https://github.com/ollama/ollama). Tested models are:

* tinyllama:latest (1b)
* llama3:latest (8b)
* qwen2:72b

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

JohannesGaessler commented:
I am one of the primary llama.cpp CUDA developers. I would in principle be willing to buy a Moore Threads GPU and to test any code changes I do in order to assert that they don't break MUSA. On the Moore Threads website I only see a "Buy Now" button for the MTT S80. Would testing and performance optimization on that GPU be representative of an MTT S4000?

yeahdongcn commented:
> I am one of the primary llama.cpp CUDA developers. I would in principle be willing to buy a Moore Threads GPU and to test any code changes I do in order to assert that they don't break MUSA. On the Moore Threads website I only see a "Buy Now" button for the MTT S80. Would testing and performance optimization on that GPU be representative of an MTT S4000?

Thank you for checking out this PR! Yes, the current code changes were tested on the MTT S4000 (--cuda-gpu-arch=mp_22) and this model of GPU only ships with our data center solution. I will test the code changes on the MTT S80 (--cuda-gpu-arch=mp_21) and let you know the results.

JohannesGaessler commented on Makefile:
```diff
@@ -598,8 +620,12 @@ endif # GGML_CUDA_DEBUG
-	NVCC = $(CCACHE) nvcc
-endif #GGML_CUDA_NVCC
+	ifdef GGML_MUSA
+		NVCC = $(CCACHE) mcc
```
Did you confirm whether the MUSA compiler actually works with ccache?

yeahdongcn replied:
Yes. I can confirm that.

```bash
root@8cb6823fed5c:/ws# make GGML_MUSA=1
I ccache found, compilation results will be cached. Disable with LLAMA_NO_CCACHE.
I llama.cpp build info: 
I UNAME_S:   Linux
I UNAME_P:   x86_64
I UNAME_M:   x86_64
I CFLAGS:    -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_MUSA -DGGML_USE_LLAMAFILE -DGGML_USE_CUDA -I/usr/local/musa/include  -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -march=native -mtune=native -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion 
I CXXFLAGS:  -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread  -march=native -mtune=native -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_MUSA -DGGML_USE_LLAMAFILE -DGGML_USE_CUDA -I/usr/local/musa/include 
I NVCCFLAGS: -std=c++11 -O3 -x musa -mtgpu --cuda-gpu-arch=mp_21 -arch=native -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -DGGML_CUDA_PEER_MAX_BATCH_SIZE=128 
I LDFLAGS:   -lmusa -lmublas -lmusart -lpthread -ldl -lrt -L/usr/local/musa/lib -L/usr/lib64 
I CC:        clang version 14.0.0 (git@sh-code.mthreads.com:sw/mtcc.git d4847756f2c727647d3956a6a7713ce1ef175967)
I CXX:       clang version 14.0.0 (git@sh-code.mthreads.com:sw/mtcc.git d4847756f2c727647d3956a6a7713ce1ef175967)
I NVCC:      InstalledDir: /usr/local/musa/bin

...

/usr/bin/ccache mcc -std=c++11 -O3 -x musa -mtgpu --cuda-gpu-arch=mp_21 -arch=native -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -DGGML_CUDA_PEER_MAX_BATCH_SIZE=128  -Iggml/include -Iggml/src -Iinclude -Isrc -Icommon -D_XOPEN_SOURCE=600 -D_GNU_SOURCE -DNDEBUG -DGGML_USE_MUSA -DGGML_USE_LLAMAFILE -DGGML_USE_CUDA -I/usr/local/musa/include  -c ggml/src/ggml-cuda/fattn-tile-f32.cu -o ggml/src/ggml-cuda/fattn-tile-f32.o
clang-14: warning: argument unused during compilation: '-arch=native' [-Wunused-command-line-argument]
In file included from ggml/src/ggml-cuda/fattn-tile-f32.cu:1:
ggml/src/ggml-cuda/common.cuh:554:1: warning: function declared 'noreturn' should not return [-Winvalid-noreturn]
}
^
ggml/src/ggml-cuda/common.cuh:658:1: warning: non-void function does not return a value [-Wreturn-type]
}
```

JohannesGaessler commented on ggml/src/ggml-cuda.cu:
```diff
@@ -1862,6 +1863,7 @@ static void ggml_cuda_mul_mat_batched_cublas(ggml_backend_cuda_context & ctx, co
                 ne23,
                 cu_compute_type,
                 CUBLAS_GEMM_DEFAULT_TENSOR_OP));
+#endif // GGML_USE_MUSA
```
If for whatever reason this code path does not work on MUSA it needs a `GGML_ASSERT(false)` instead of just silently failing.

yeahdongcn replied:
Thanks for pointing out the issue. I will refine it and update the PR.

yeahdongcn replied:
Please see the latest commits. Thanks.

JohannesGaessler commented on ggml/src/ggml-cuda.cu:
```diff
@@ -1884,9 +1886,9 @@ static void ggml_cuda_mul_mat(ggml_backend_cuda_context & ctx, const ggml_tensor
+#if !defined(GGML_CUDA_FORCE_DMMV) && !defined(GGML_USE_MUSA)
     use_dequantize_mul_mat_vec = use_dequantize_mul_mat_vec && !use_mul_mat_vec_q;
-#endif // GGML_CUDA_FORCE_DMMV
+#endif // !GGML_CUDA_FORCE_DMMV && !GGML_USE_MUSA
```
Why are you making changes to the `ggml_cuda_mul_mat` logic? If the issue has to do with the availability of `__dp4a` it would be better to adjust `ggml_cuda_dp4a` in `common.cuh`.

yeahdongcn replied:
This is because in https://github.com/ggerganov/llama.cpp/commit/cb5fad4c6c2cbef92e9b8b63449e1cb7664e4846, `use_dequantize_mul_mat_vec` was moved upper and no longer checks against `cc >= MIN_CC_DP4A`. As a result, the `ggml_cuda_cpy_tensor_2d` function returns an error when executed on the MUSA device. Although I haven't identified the root cause, I reverted the code to its previous working version as a temporary solution.

yeahdongcn replied:
I just updated the code to make it more explicit. Please see the latest commit. Thanks.

JohannesGaessler commented on ggml/src/ggml-cuda/common.cuh:
```diff
@@ -114,6 +117,151 @@
+#define cublasComputeType_t cudaDataType_t
+
+// XXX: Clang builtins mapping
+#define __vsubss4 __vsubss4_musa
+#define __vsub4   __vsub4_musa
+#define __vcmpeq4 __vcmpeq4_musa
+#define __vcmpne4 __vcmpne4_musa
```
These HIP workarounds that were taken over for MUSA have comparatively poor performance. Make sure to check whether there are equivalent instructions that could be used instead.

yeahdongcn replied:
Yes. This is because our compiler has declared the builtins but not implemented at the moment. Thus, I chose to copy HIP code and rename to something like `__vsubss4_musa` and etc. Then use macro to replace the original `__vsubss4` to `__vsubss4_musa`. Once the compiler is ready, I can remove the macro and use the original `__vsubss4` directly.

JohannesGaessler commented on src/llama.cpp:
```diff
@@ -18733,7 +18733,7 @@ struct llama_model_params llama_model_default_params() {
     };
 
-#ifdef GGML_USE_METAL
+#if defined(GGML_USE_METAL) || defined(GGML_USE_MUSA)
```
I don't think this is necessary. If MUSA works the same way as HIP then CPU+GPU hybrid inference should work without the need for any additional changes.

yeahdongcn replied:
Yes. This is quite weird. By using Ollama, the model could be offloaded to the GPU. While using llama-server directly, I saw this log: `llm_load_tensors: offloaded 0/33 layers to GPU`. So I marked `result.n_gpu_layers = 999` the same as metal.

slaren replied:
This makes sense with Metal because it uses the same memory as the CPU, so there is rarely any reason to not fully offload the model. Problems with ollama should be fixed in ollama.

JohannesGaessler replied:
>While using llama-server directly, I saw this log: llm_load_tensors: offloaded 0/33 layers to GPU. 

Did you set the `--n-gpu-layers` CLI argument?

yeahdongcn replied:
No. I just run llama-server like:

```bash
root@8cb6823fed5c:/ws# ./llama-server
INFO [                    main] build info | tid="139686874033984" timestamp=1720524760 build=3331 commit="777add8b"
INFO [                    main] system info | tid="139686874033984" timestamp=1720524760 n_threads=6 n_threads_batch=-1 total_threads=12 system_info="AVX = 1 | AVX_VNNI = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | AVX512_BF16 = 0 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 | "
llama_model_loader: loaded meta data with 23 key-value pairs and 201 tensors from models/7B/ggml-model-f16.gguf (version GGUF V3 (latest))
...
ggml_cuda_init: found 1 MUSA devices:
  Device 0: MTT S4000, compute capability 2.2, VMM: yes
llm_load_tensors: ggml ctx size =    0.27 MiB
llm_load_tensors: offloading 0 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 0/33 layers to GPU
...
```

JohannesGaessler replied:
The default value for `--n-gpu-layers` is 0 so you are effectively not using the GPU at all.

yeahdongcn replied:
Thanks for letting me know this. I will revert this change.

yeahdongcn replied:
Please see the latest commits. Thanks.

yeahdongcn commented on Makefile:
I should refine this. But my initial thought is to mostly reuse the CUDA compiler settings and only change the different parts. This is because the CUDA compiler is already well-configured and I don't want to mess up with it.

yeahdongcn replied:
After conducting some local testing, I found that if we separate out the MUSA changes from the CUDA logic in the Makefile/CMakeLists.txt, we would end up duplicating a significant amount of code. This would lead to maintenance issues down the line.
I'd love to explore alternative solutions but I'm not sure what the best way to proceed is. I'm open to suggestions.

JohannesGaessler replied:
As I said, I don't feel particularly strongly about this. But what I would like is a consistent implementation between HIP and MUSA. So if we take the original approach of this PR then we should also adjust the HIP logic (in a separate PR).

yeahdongcn replied:
Ok, I understand.

slaren commented on Makefile:
```diff
@@ -526,6 +526,14 @@ ifndef GGML_NO_ACCELERATE
+	CC := clang
+	CXX := clang++
+	GGML_CUDA := 1
+	GGML_NO_OPENMP := 1
```
I assume that the issue with OpenMP is that the MUSA compiler does not work well it. Could the MUSA compiler be used only for the ggml-cuda files instead of completely disabling OpenMP?

yeahdongcn replied:
Thank you for pointing this out! The MUSA compiler is shipped with our Clang variant, which is based on Clang 14.

1. Ubuntu 20.04 Requirement: MUSA libraries require Ubuntu 20.04.
2. OpenMP Version: The latest OpenMP version available on Ubuntu 20.04 (Focal) is `libomp-12-dev`.

I'm not sure if setting `GGML_NO_OPENMP` to `0` will cause runtime issues. Therefore, I have set `GML_NO_OPENMP` to `1` to disable it. Once we upgrade to Ubuntu 22.04, we can re-enable it.

yeahdongcn replied:
BTW, just saw this in my previous work log:

```bash
root@b7a6e65eec34:/ws# ldd llama-server 
        linux-vdso.so.1 (0x00007ffdff2de000)
        libmusa.so.1.0 => /usr/local/musa/lib/libmusa.so.1.0 (0x00007f044742e000)
        libmublas.so => /usr/local/musa/lib/libmublas.so (0x00007f043c1a0000)
        libmusart.so.1.0 => /usr/local/musa/lib/libmusart.so.1.0 (0x00007f043c16e000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f043c144000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f043c13e000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f043c134000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f043bf50000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f043be01000)
        libomp.so.5 => /lib/x86_64-linux-gnu/libomp.so.5 (0x00007f043bcff000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f043bce4000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f043baf2000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f04477a9000)
        libelf.so.1 => /lib/x86_64-linux-gnu/libelf.so.1 (0x00007f043bad6000)
        libsrv_um_MUSA.so => /ddk/usr/lib/x86_64-linux-gnu/musa/libsrv_um_MUSA.so (0x00007f043b1bf000)
        libnuma.so.1 => /lib/x86_64-linux-gnu/libnuma.so.1 (0x00007f043b1b2000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f043b196000)
```

slaren replied:
It will work without OpenMP, but the performance will be significantly worse in some scenarios such as `-nkvo`.

yeahdongcn replied:
Yes, I understand. I can perform some tests in the runtime env and check if we can enable it in this PR.

yeahdongcn replied:
I just want to confirm: will llama.cpp accept a hardcoded path like this?

```diff
diff --git a/Makefile b/Makefile
index d2d9ce0f..4d10bd74 100644
--- a/Makefile
+++ b/Makefile
@@ -530,7 +530,6 @@ ifdef GGML_MUSA
        CC := clang
        CXX := clang++
        GGML_CUDA := 1
-       GGML_NO_OPENMP := 1
        MK_CPPFLAGS += -DGGML_USE_MUSA
 endif
 
@@ -589,8 +588,8 @@ ifdef GGML_CUDA
                        CUDA_PATH ?= /usr/local/musa
                endif
 
-               MK_CPPFLAGS  += -DGGML_USE_CUDA -I$(CUDA_PATH)/include
-               MK_LDFLAGS   += -lmusa -lmublas -lmusart -lpthread -ldl -lrt -L$(CUDA_PATH)/lib -L/usr/lib64
+               MK_CPPFLAGS  += -DGGML_USE_CUDA -I$(CUDA_PATH)/include -I/usr/lib/llvm-12/lib/clang/12.0.0/include
+               MK_LDFLAGS   += -lmusa -lmublas -lmusart -lpthread -ldl -lrt -L$(CUDA_PATH)/lib -L/usr/lib64 -L/usr/lib/llvm-12/lib
                MK_NVCCFLAGS += -x musa -mtgpu --cuda-gpu-arch=mp_22
        else
                ifneq ('', '$(wildcard /opt/cuda)')
```

slaren replied:
If that's always the path for the MUSA compiler that may be ok, it depends entirely on the compiler, not on llama.cpp. If it is not, it may cause issues when other people try to build with MUSA.

yeahdongcn replied:
I switched to the default `libomp-dev` in the builder image and `libomp5-10` in the runtime image.
Now `GGML_NO_OPENMP := 1` is removed from Makefile (as well as `set(GGML_OPENMP OFF)` in CMake) and everything looks good in my local tests.
Please see the latest commits. Thanks.

slaren replied:
I see that the `.cu` files are being compiled with the `mcc` compiler, so why change the C/C++ compiler to `clang` for the rest of the project?

yeahdongcn replied:
`mcc` is actually a link to `clang`.

```bash
root@9627f6e94585:/ws# ll /usr/local/musa/bin/mcc
lrwxrwxrwx 1 root root 5 Jul 11 13:16 /usr/local/musa/bin/mcc -> clang*
```

The version of `clang` ships with MUSA, contains some extra modifications.

yeahdongcn committed:
Update doc for MUSA

Signed-off-by: Xiaodong Ye <xiaodong.ye@mthreads.com>

slaren commented on ggml/src/CMakeLists.txt:
```diff
@@ -1059,7 +1108,9 @@ if (GGML_CUDA)
 
     if (NOT CUDA_CXX_FLAGS_JOINED STREQUAL "")
-        list(APPEND CUDA_FLAGS -Xcompiler ${CUDA_CXX_FLAGS_JOINED})
+        # list(APPEND CUDA_FLAGS -Xcompiler ${CUDA_CXX_FLAGS_JOINED})
+        # XXX: Removed flags: -Xcompiler
+        list(APPEND CUDA_FLAGS ${CUDA_CXX_FLAGS_JOINED})
```
`-Xcompiler` is necessary for old versions of the CUDA toolkit.

yeahdongcn replied:
This was accidental. I have reverted the change.

slaren commented on ggml/src/CMakeLists.txt:
```diff
@@ -139,6 +139,19 @@ if (GGML_METAL)
+
+    set(GGML_CUDA ON)
+
+    list(APPEND GGML_CDEF_PUBLIC GGML_USE_MUSA)
+
+    add_compile_definitions(GGML_USE_MUSA)
```
The `add_compile_definitions` is redundant here, adding to `GGML_CDEF_PUBLIC` is enough.

slaren commented on ggml/src/ggml-cuda.cu:
```diff
@@ -1902,11 +1906,17 @@ static void ggml_cuda_mul_mat(ggml_backend_cuda_context & ctx, const ggml_tensor
             const int cc            = ggml_cuda_info().devices[id].cc;
             use_mul_mat_q           = use_mul_mat_q           && ggml_cuda_should_use_mmq(src0->type, cc, src1->ne[1]);
             any_gpus_with_slow_fp16 = any_gpus_with_slow_fp16 || !fast_fp16_available(cc);
+#ifdef GGML_USE_MUSA
+            use_mul_mat_vec_q       = false;
+#endif // GGML_USE_MUSA
         }
     } else {
         const int cc            = ggml_cuda_info().devices[ctx.device].cc;
         use_mul_mat_q           = use_mul_mat_q           && ggml_cuda_should_use_mmq(src0->type, cc, src1->ne[1]);
         any_gpus_with_slow_fp16 = any_gpus_with_slow_fp16 || !fast_fp16_available(cc);
+#ifdef GGML_USE_MUSA
+        use_mul_mat_vec_q       = false;
+#endif // GGML_USE_MUSA
```
This is going to break types such as IQ quants that are not supported by dmmv.

yeahdongcn replied:
Yes, we are experiencing issues with enabling `use_mul_mat_vec_q`. For now, I have to explicitly disable it.

yeahdongcn replied:
After further investigation, it appears that the issue is linked to `vmm` enablement. To resolve the problem temporarily, I have disabled `vmm` and rolled back my previous modifications to the `use_mul_mat_vec_q` code.

yeahdongcn committed:
Add GGML_MUSA in Makefile

Signed-off-by: Xiaodong Ye <xiaodong.ye@mthreads.com>

yeahdongcn committed:
Add GGML_MUSA in CMake

Signed-off-by: Xiaodong Ye <xiaodong.ye@mthreads.com>

yeahdongcn committed:
CUDA => MUSA

Signed-off-by: Xiaodong Ye <xiaodong.ye@mthreads.com>

yeahdongcn commented:
@JohannesGaessler @slaren I've addressed most of your comments—thank you again for the review. However, two comments related to compilation remain unresolved. I am currently collaborating with our compiler team to address these issues, but it may take longer than anticipated. Are there any other concerns regarding the remaining changes?

yeahdongcn committed:
MUSA adds support for __vsubss4

Signed-off-by: Xiaodong Ye <xiaodong.ye@mthreads.com>

slaren reviewed (approved):
Eventually we should move all the HIP and MUSA-specific code to its own headers.

yeahdongcn commented:
> Eventually we should move all the HIP and MUSA-specific code to its own headers.

No problem. I can start working on this.

yeahdongcn committed:
Fix CI build failure

Signed-off-by: Xiaodong Ye <xiaodong.ye@mthreads.com>

yeahdongcn commented:
> In an earlier post you said:
> 
> > Thank you for checking out this PR! Yes, the current code changes were tested on the MTT S4000 (--cuda-gpu-arch=mp_22) and this model of GPU only ships with our data center solution. I will test the code changes on the MTT S80 (--cuda-gpu-arch=mp_21) and let you know the results.
> 
> Have there been any updates on this?

I've encountered some compilation issues on S80 toolchain and have opened several internal tickets to the compiler team. I'll monitor the progress and keep you updated.

1823616178 commented:
make error 

ubuntu 20.04.06 LTS
musa driver 2.7.0 
SDK: MUSA+SDK-rc2.1.0_Intel_CPU_Ubuntu_chunxiao

make GGML_MUSA=1
![image](https://github.com/user-attachments/assets/3be767e4-1c90-49de-ab2c-151ed55d4150)
Please help me.

1823616178 commented:
My video card is s80,cpu is amd 2600

yeahdongcn commented:
> make error
> 
> ubuntu 20.04.06 LTS musa driver 2.7.0 SDK: MUSA+SDK-rc2.1.0_Intel_CPU_Ubuntu_chunxiao
> 
> make GGML_MUSA=1 ![image](https://private-user-images.githubusercontent.com/24314110/352893441-3be767e4-1c90-49de-ab2c-151ed55d4150.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyMjA2MzEsIm5iZiI6MTcyMjIyMDMzMSwicGF0aCI6Ii8yNDMxNDExMC8zNTI4OTM0NDEtM2JlNzY3ZTQtMWM5MC00OWRlLWFiMmMtMTUxZWQ1NWQ0MTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDAyMzIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjMTg0M2MzYzRmZWVmNDU2M2QxNTMyODhhYmExZDkxYzIxYjIwY2JkZmQ3YzNiMzgxZDViMGMwN2Q5MWUzZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V6TrPByy6H4DNEV8JGWc-RzEvkEQk1EpLRvRtDwm5Ww) Please help me.

Please see the above comments:

> I've encountered some compilation issues on S80 toolchain and have opened several internal tickets to the compiler team. I'll monitor the progress and keep you updated.
> 
> The S80 toolchain (rc2.1.0_Intel_CPU_Ubuntu_quyuan) I used is publicly available but still in the RC stage. Please refer to [link](https://developer.mthreads.com/sdk/download/musa?equipment=&os=&driverVersion=&version=)

We are still investigating this issue internally. Please expect a new release of MUSA SDK and llama.cpp PR.

XenoAmess commented:
> > make error
> > ubuntu 20.04.06 LTS musa driver 2.7.0 SDK: MUSA+SDK-rc2.1.0_Intel_CPU_Ubuntu_chunxiao
> > make GGML_MUSA=1 ![image](https://private-user-images.githubusercontent.com/24314110/352893441-3be767e4-1c90-49de-ab2c-151ed55d4150.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyMjA2MzEsIm5iZiI6MTcyMjIyMDMzMSwicGF0aCI6Ii8yNDMxNDExMC8zNTI4OTM0NDEtM2JlNzY3ZTQtMWM5MC00OWRlLWFiMmMtMTUxZWQ1NWQ0MTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDAyMzIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjMTg0M2MzYzRmZWVmNDU2M2QxNTMyODhhYmExZDkxYzIxYjIwY2JkZmQ3YzNiMzgxZDViMGMwN2Q5MWUzZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V6TrPByy6H4DNEV8JGWc-RzEvkEQk1EpLRvRtDwm5Ww) Please help me.
> 
> Please see the above comments:
> 
> > I've encountered some compilation issues on S80 toolchain and have opened several internal tickets to the compiler team. I'll monitor the progress and keep you updated.
> > The S80 toolchain (rc2.1.0_Intel_CPU_Ubuntu_quyuan) I used is publicly available but still in the RC stage. Please refer to [link](https://developer.mthreads.com/sdk/download/musa?equipment=&os=&driverVersion=&version=)
> 
> We are still investigating this issue internally. Please expect a new release of MUSA SDK and llama.cpp PR.

any progress in s80?

1823616178 commented:
> > > make error
> > > ubuntu 20.04.06 LTS musa driver 2.7.0 SDK: MUSA+SDK-rc2.1.0_Intel_CPU_Ubuntu_chunxiao
> > > make GGML_MUSA=1 ![image](https://private-user-images.githubusercontent.com/24314110/352893441-3be767e4-1c90-49de-ab2c-151ed55d4150.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyMjA2MzEsIm5iZiI6MTcyMjIyMDMzMSwicGF0aCI6Ii8yNDMxNDExMC8zNTI4OTM0NDEtM2JlNzY3ZTQtMWM5MC00OWRlLWFiMmMtMTUxZWQ1NWQ0MTUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDAyMzIxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjMTg0M2MzYzRmZWVmNDU2M2QxNTMyODhhYmExZDkxYzIxYjIwY2JkZmQ3YzNiMzgxZDViMGMwN2Q5MWUzZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V6TrPByy6H4DNEV8JGWc-RzEvkEQk1EpLRvRtDwm5Ww) Please help me.
> > 
> > 
> > Please see the above comments:
> > > I've encountered some compilation issues on S80 toolchain and have opened several internal tickets to the compiler team. I'll monitor the progress and keep you updated.
> > > The S80 toolchain (rc2.1.0_Intel_CPU_Ubuntu_quyuan) I used is publicly available but still in the RC stage. Please refer to [link](https://developer.mthreads.com/sdk/download/musa?equipment=&os=&driverVersion=&version=)
> > 
> > 
> > We are still investigating this issue internally. Please expect a new release of MUSA SDK and llama.cpp PR.
> 
> any progress in s80?

I guess we'll have to wait for the next version of the SDK

yeahdongcn commented:
> I guess we'll have to wait for the next version of the SDK

Yes, please give us more time.

Ivening commented:
@yeahdongcn please help - have got the problem with compiling llama.cpp using MUSA SDK rc2.0.0 with Ubuntu 20.04.6 LTS:
running `make GGML_MUSA=1` shows the following error:
![Screenshot 2024-09-19 141040](https://github.com/user-attachments/assets/5a0aa59d-1e15-48ac-b524-23ea6994bd81)
Is there something I am doing wrong for the compilation?

Ivening commented:
@yeahdongcn thank you for your reply! Will this code work with MTT S3000?

yeahdongcn commented:
> @yeahdongcn thank you for your reply! Will this code work with MTT S3000?

Haha, it seems that you're one of our business customers! MTT S3000 shares the same architecture as MTT S80, I can test on MTT S3000 as well.

arch-btw commented:
@yeahdongcn what speeds can we expect for ~8B models for the MTT S80?

yeahdongcn commented:
> @yeahdongcn what speeds can we expect for ~8B models for the MTT S80?

~15 tokens/s (llama3.1:8b)

arch-btw commented:
@yeahdongcn very good, thank you
